### PR TITLE
Cache http

### PR DESCRIPTION
**Pull request by scottgifford**
https://issues.apache.org/jira/projects/HTTPCORE/issues/HTTPCORE-615

### CODE REVIEW

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/cache/ResourceIOException.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
     private static final long serialVersionUID = 1L;
 
     public ResourceIOException(final String message) {
-        super();
+        super(message);
```
Comment (michael-o) [inline httpclient5-cache/src/main/java/org/apache/hc/client5/http/cache/ResourceIOException.java]: That's weird. Looks like a bug..

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+    /**
+     * Cache validity policy that always returns an age of 0.
+     *
+     * This prevents the Age header from being written to the cache (it does not make sense to cache it),
+     * and is the only thing the policy is used for in this case.
+     */
+    private class NoAgeCacheValidityPolicy extends CacheValidityPolicy {
+        @Override
+        public long getCurrentAgeSecs(final HttpCacheEntry entry, final Date now) {
```
Comment (michael-o) [inline httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java]: Please note that will soon move to `TimeValue`.

**Comment by ok2c** [comment]
@scottgifford Could you please squash your changes into a single commit (or a few logically self-contained commits)?

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/cache/ResourceIOException.java (hunk 1 above)
Comment (scottgifford) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/cache/ResourceIOException.java]: Yeah agreed.

**Comment by scottgifford** [comment]
> I am somewhat confused why the memcached is competely gone. The approach seems to be the same.
> Logically, it looks fine to me, but I am not acquant with the cache code at all.

It is only a change in naming, following a corresponding change in naming and approach between the 4.5.x branch and master.

In the 4.5.x branch, the serializers are specific to the cache; the interface implemented for 4.5.x was called [MemcachedCacheEntry](https://github.com/apache/httpcomponents-client/blob/4.5.x/httpclient-cache/src/main/java/org/apache/http/impl/client/cache/memcached/MemcachedCacheEntry.java) and the reference implementation is [MemcachedCacheEntryImpl](https://github.com/apache/httpcomponents-client/blob/4.5.x/httpclient-cache/src/main/java/org/apache/http/impl/client/cache/memcached/MemcachedCacheEntryImpl.java).  So my original change followed that convention.

In the master branch, the serializers are not cache-specific. and the naming convention has been changed to reflect that.  Here the interface I am implementing is called [HttpCacheEntrySerializer](https://github.com/apache/httpcomponents-client/blob/master/httpclient5-cache/src/main/java/org/apache/hc/client5/http/cache/HttpCacheEntrySerializer.java) and the reference implementation is [ByteArrayCacheEntrySerializer](https://github.com/apache/httpcomponents-client/blob/master/httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/ByteArrayCacheEntrySerializer.java).  My updated code follows this convention.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java (hunk 2 above)
Comment (scottgifford) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java]: OK, seems like it should be an easy change, but not one I can make quite yet.

**Comment by scottgifford** [comment]
> @scottgifford Could you please squash your changes into a single commit (or a few logically self-contained commits)?

Sure will do.

**Comment by scottgifford** [comment]
> > @scottgifford Could you please squash your changes into a single commit (or a few logically self-contained commits)?
> 
> Sure will do.

Squashed commits are in PR https://github.com/apache/httpcomponents-client/pull/192

**Comment by scottgifford** [comment]
Closed in favor of https://github.com/apache/httpcomponents-client/pull/192

**Pull request closed without merging**